src/world/tree_generation.js

```diff
@@ -1,10 +1,107 @@
 import { alea } from "../libs/alea/alea.js";
 import { rangeFrom, rangeList } from "../utils.js";
 import { SeedFork } from "./seed.js";
+import { BaseGenerator } from "./base_generator.js";
 
 
+export class TreePlacer extends BaseGenerator {
+  constructor(game) {
+    super(game);
+  }
+
+  getGenerator() {
+    switch (this.gcnf.treeCollideAction) {
+      case "avoid":
+        return new AvoidTreePlacer(
+          this.globSeed, this.wSize[0], this.wSize[2], 
+          this.gcnf.nTrees, this.gcnf.treeRadius);
+      case "place":
+        return new IgnoreTreePlacer(this);
+      case "skip":
+        return new SkipTreePlacer(this);
+      default:
+        throw new ReferenceError("Unknown treePlaceMode");
+    }
+  }
+
+  makeTrees() {
+    return this.getGenerator().makeTrees();
+  }
+}
+
+
+export class IgnoreTreePlacer extends BaseGenerator {
+  constructor(game) {
+    super(game);
+    this.n = this.wSize[0] * this.wSize[2];
+    this.seed = this.getSeed("tree-pos", 0);
+    this.rng = alea(this.seed);
+  }
+
+  makeTrees() {
+    return rangeList(this.n).map(_ => this.idxToCoord(this.rng.randint(0, this.n)));
+  }
+  /**
+   * @param {number} idx
+   * @retuns {[number, number]}
+   */
+  idxToCoord(idx) {
+    let z = Math.floor(idx / this.wSize[0]);
+    let x = idx % this.wSize[0];
+    return [x, z];
+  }
+}
+
+
+export class SkipTreePlacer extends BaseGenerator {
+  constructor(game) {
+    super(game);
+    this.n = this.wSize[0] * this.wSize[2];
+    this.seed = this.getSeed("tree-pos", 0);
+    this.rng = alea(this.seed);
+    let w = this.gcnf.treeRadius[0];
+    let h = this.gcnf.treeRadius[1];
+    /** @type {Array<[number, number]>} */
+    this.excludeOffsets = rangeFrom(-w, w+1)
+      .flatMap(x => rangeFrom(-h, h+1)
+        .map(y => [x, y]));
+  }
+
+  makeTrees() {
+    let columns = new Uint8Array(this.n);
+    return rangeList(this.n).map(_ => {
+      let idx = this.rng.randint(0, this.n);
+      let c = this.idxToCoord(idx);
+      if(!columns[idx]) {
+        for(const [xo, yo] of this.excludeOffsets) {
+          columns[this.coordToIdx(c[0] + xo, c[1] + yo)] = 1;
+        }
+        return c;
+      }
+    }).filter(c => c != null);
+  }
+  /**
+   * @param {number} idx
+   * @retuns {[number, number]}
+   */
+  idxToCoord(idx) {
+    let z = Math.floor(idx / this.wSize[0]);
+    let x = idx % this.wSize[0];
+    return [x, z];
+  }
+
+  /**
+   * @param {number} x
+   * @param {number} z
+   * @returns {number}
+   */
+  coordToIdx(x, z) {
+    return this.wSize[0] * z + x;
+  }
+}
+
 
-export class TreePosGetter {
+export class AvoidTreePlacer {
   constructor(seed, x, z, n_trees, treeRadius) {
```

Comment:
Does this really need this many arguments? It would be cleaner just to pass the `Game` object (or the `TreePlacer` object) and extract the needed attributes from that object.